src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2288,9 +2288,15 @@ public override BoundNode VisitDefaultExpression(BoundDefaultExpression node)
             Debug.Assert(!IsConditionalState);
             if (this.State.Reachable)
             {
-                if ((object)node.Type != null && node.Type.IsReferenceType == true)
+                if ((object)node.Type != null && node.Type.IsReferenceType)
                 {
-                    this.State.ResultIsNotNull = false;
+                    bool isNotNull = node.Syntax.Kind() == SyntaxKind.DefaultExpression && !node.IsNullable;
```

Comment:
>node.Syntax.Kind() == SyntaxKind.DefaultExpression [](start = 37, length = 50)

I would prefer us to not rely on syntax, but make sure that IsNullable is properly set by the binder.